### PR TITLE
Implemented Custom role and permissions

### PR DESCRIPTION
**Pull request by addisonbeck**
https://app.asana.com/0/1195018406457675/1153292148278664
bitwarden/jslib#237
bitwarden/web#750
bitwarden/browser/pull/1505
bitwarden/cli/pull/204
bitwarden/mobile/pull/1189

### Narrative
As a user, I would like to be able to configure another user with access to manage user/group/collection assignment without being able to see organization passwords that are not directly assigned to them.

### Changes
1. Added a new string column to [dbo].[OrganizationUser] called Permissions to store a JSON blob of user permissions .
2. Added a new role type: `Custom`
3. Added new potential claims for each created permission & the custom role type
4. Created new `CurrentContext` getters for calculating if the logged in user can perform a permission-blocked operation that checks their role and the new `OrganizationUser` columns.
5. Replaced any relevant calls to `isAdmin()`/`isManager()`/`isOwner()` with calls to check their specific permissions.
6. Created a new `OrganizationInvite.cs` model to slim down the `OrganizationService`'s invite method parameters, and removed one of the invite methods that could be easily combined with another.
7. Gave the `OrganizationService` a private method used to authorize actions based on the role/permissions hierarchy & implemented in places that were already performing those checks individually.

### CODE REVIEW

File: bitwarden_license/src/Portal/Controllers/SsoController.cs [hunk 1]
```diff
@@ -41,7 +41,7 @@ public async Task<IActionResult> Index()
             }
 
             if (!_enterprisePortalCurrentContext.SelectedOrganizationDetails.UseSso ||
-                !_enterprisePortalCurrentContext.AdminForSelectedOrganization)
+                !_enterprisePortalCurrentContext.CanManagePoliciesForSelectedOrganization)
```
Comment (kspearrin) [inline bitwarden_license/src/Portal/Controllers/SsoController.cs:44]: Why can they edit SSo if they have policy permissions? Seems unrelated?

File: src/Api/Public/Controllers/MembersController.cs [hunk 2]
```diff
@@ -116,8 +117,15 @@ public async Task<IActionResult> List()
         public async Task<IActionResult> Post([FromBody]MemberCreateRequestModel model)
         {
             var associations = model.Collections?.Select(c => c.ToSelectionReadOnly());
-            var user = await _organizationService.InviteUserAsync(_currentContext.OrganizationId.Value, null,
-                model.Email, model.Type.Value, model.AccessAll.Value, model.ExternalId, associations);
+            var invite = new OrganizationUserInvite()
```
Comment (cscharf) [inline src/Api/Public/Controllers/MembersController.cs:120]: Remove empty parens on simplified object initialization.

File: src/Core/CurrentContext.cs [hunk 3]
```diff
@@ -148,6 +148,44 @@ public virtual Task SetContextAsync(ClaimsPrincipal user)
...
             }
+
+            if (claimsDict.ContainsKey("orgcustom"))
+            {
+                Organizations.AddRange(claimsDict["orgcustom"].Select(c =>
+                    new CurrentContentOrganization
+                    {
+                        Id = new Guid(c.Value),
+                        Type = OrganizationUserType.Custom,
+                        AccessBusinessPortal = claimsDict.ContainsKey("accessbusinessportal") ?
```
Comment (cscharf) [inline src/Core/CurrentContext.cs:159]: This code smells like a new private method, anonymous method or extension method to me:

```c#
claimsDict.ContainsKey("accessbusinessportal") ? 
   claimsDict["accessbusinessportal"].Any(x => x.Value == c.Value) : false
```

File: src/Core/Enums/OrganizationUserType.cs [hunk 4]
```diff
@@ -6,5 +6,6 @@ public enum OrganizationUserType : byte
         Admin = 1,
         User = 2,
         Manager = 3,
+        Custom = 4
```
Comment (cscharf) [inline src/Core/Enums/OrganizationUserType.cs:9]: Please end all enums with a trailing comma, so when we add new ones (like you're doing here) we don't have to also modify the line above it.

File: src/Core/IdentityServer/ApiResources.cs [hunk 5]
```diff
@@ -20,7 +20,8 @@ public static IEnumerable<ApiResource> GetApiResources()
                     "orgowner",
                     "orgadmin",
                     "orgmanager",
-                    "orguser"
+                    "orguser",
+                    "orgcustom"
```
Comment (cscharf) [inline src/Core/IdentityServer/ApiResources.cs:24]: Trailing comma at the end of the list here please (same reason as the enums above).

File: src/Core/Models/Api/Request/Organizations/OrganizationUpdateRequestModel.cs [hunk 6]
```diff
@@ -1,9 +1,10 @@
-﻿using Bit.Core.Models.Table;
+﻿using Bit.Core.Models.Interfaces;
+using Bit.Core.Models.Table;
 using System.ComponentModel.DataAnnotations;
 
 namespace Bit.Core.Models.Api
 {
-    public class OrganizationUpdateRequestModel
+    public class OrganizationUpdateRequestModel: IPermissions
```
Comment (cscharf) [inline src/Core/Models/Api/Request/Organizations/OrganizationUpdateRequestModel.cs:7]: Space before the colon, ` : `

File: src/Core/Models/Business/OrganizationUserInvite.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
+using System.Collections.Generic;
+using Bit.Core.Models.Api;
+using Bit.Core.Models.Interfaces;
+using System.Linq;
+using Bit.Core.Models.Data;
+
+namespace Bit.Core.Models.Business
+{
+    public class OrganizationUserInvite: IPermissions
```
Comment (cscharf) [inline src/Core/Models/Business/OrganizationUserInvite.cs:9]: Spacing before the colon, ` : `

File: src/Core/Models/Business/OrganizationUserInvite.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
+using System.Collections.Generic;
+using Bit.Core.Models.Api;
+using Bit.Core.Models.Interfaces;
+using System.Linq;
+using Bit.Core.Models.Data;
```
Comment (cscharf) [inline src/Core/Models/Business/OrganizationUserInvite.cs:5]: Please consider sorting your using directives.

File: src/Core/Models/Data/OrganizationUserOrganizationDetails.cs [hunk 9]
```diff
@@ -1,8 +1,9 @@
 ﻿using System;
+using Bit.Core.Models.Interfaces;
 
 namespace Bit.Core.Models.Data
 {
-    public class OrganizationUserOrganizationDetails
+    public class OrganizationUserOrganizationDetails: IPermissions
```
Comment (cscharf) [inline src/Core/Models/Data/OrganizationUserOrganizationDetails.cs:6]: Need more space around your colon 😁 , ` : `

File: src/Core/Models/Interfaces/IPermissions.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+namespace Bit.Core.Models.Interfaces 
+{
+    public interface IPermissions
+    {
+        public bool AccessBusinessPortal { get; set; }
```
Comment (cscharf) [inline src/Core/Models/Interfaces/IPermissions.cs:5]: Remove `public` keywords from interface... all members (properties/methods) of an interface are already `public`, the compiler should be yelling at you here, or at least the optimizer/intellisense.

File: src/Core/Models/Table/OrganizationUser.cs [hunk 11]
```diff
@@ -17,6 +18,15 @@ public class OrganizationUser : ITableObject<Guid>, IExternal
         public string ExternalId { get; set; }
         public DateTime CreationDate { get; internal set; } = DateTime.UtcNow;
         public DateTime RevisionDate { get; internal set; } = DateTime.UtcNow;
+        public bool AccessBusinessPortal { get; set; }
```
Comment (cscharf) [inline src/Core/Models/Table/OrganizationUser.cs:21]: Question (don't be cross), but to "future-proof" this a bit, like we did with cipher, SSO config, etc. wouldn't it behoove us to create a class to represent these permissions and just store the serialized JSON in the DB in a single column, so the update/save/insert stored procedures never need to change, and the table never needs to change, and we can just add new permissions in the future w/o mucking with the database "stuff"?

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 12]
```diff
@@ -950,45 +950,25 @@ public async Task DisableTwoFactorProviderAsync(Organization organization, TwoFa
...
         public async Task<List<OrganizationUser>> InviteUserAsync(Guid organizationId, Guid? invitingUserId,
-            IEnumerable<string> emails, OrganizationUserType type, bool accessAll, string externalId,
-            IEnumerable<SelectionReadOnly> collections)
+            string externalId, OrganizationUserInvite invite)
         {
             var organization = await GetOrgById(organizationId);
             if (organization == null)
             {
                 throw new NotFoundException();
             }
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:961]: I would add some validation around `invite` to ensure it's not `null`, etc. There are assumptions made here, and perhaps they are the correct assumptions, but I'm not tracking upstream to see if being defensive here and validating not-null makes sense (let me know if it doesn't and that's cool with me).

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 13]
```diff
@@ -1480,8 +1466,16 @@ public async Task<OrganizationLicense> GenerateLicenseAsync(Guid organizationId,
 
                         try
                         {
-                            var newUser = await InviteUserAsync(organizationId, importingUserId, user.Email,
-                                OrganizationUserType.User, false, user.ExternalId, new List<SelectionReadOnly>());
+                            var invite = new OrganizationUserInvite()
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1469]: remove empty parens when using simple object initialization

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 14]
```diff
@@ -1651,5 +1645,68 @@ private void ValidateOrganizationUpgradeParameters(Models.StaticStore.Plan plan,
...
             }
         }
+
+        private async Task ValidateOrganizationUserUpdatePermissions(Guid loggedInUserId, Guid organizationId, OrganizationUserType newType, OrganizationUserType? oldType) 
+        {
+            var checkedTypes = new List<OrganizationUserType> 
+            { 
+                OrganizationUserType.Owner,
+                OrganizationUserType.Admin,
+                OrganizationUserType.Custom
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1655]: Add trailing comma at end of list.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 15]
```diff
@@ -1480,8 +1466,16 @@ public async Task<OrganizationLicense> GenerateLicenseAsync(Guid organizationId,
...
                         try
                         {
-                            var newUser = await InviteUserAsync(organizationId, importingUserId, user.Email,
-                                OrganizationUserType.User, false, user.ExternalId, new List<SelectionReadOnly>());
+                            var invite = new OrganizationUserInvite() 
+                            {
+                                Emails = new List<string> { user.Email },
+                                Type = OrganizationUserType.User,
+                                AccessAll = false,
+                                Collections = new List<SelectionReadOnly>()
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1474]: Add trailing comma at end of list.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 16]
```diff
@@ -1651,5 +1645,68 @@ private void ValidateOrganizationUpgradeParameters(Models.StaticStore.Plan plan,
...
+                        {
+                            throw new BadRequestException("Only Owners and Admins can configure Custom accounts.");
+                        }
+
+                        var loggedInAsOrgCustomOrgUser = loggedInUserOrgs.FirstOrDefault(u => u.OrganizationId == organizationId && u.Type == OrganizationUserType.Custom);
+                        if (loggedInAsOrgCustomOrgUser != null)
+                        {
+                            if (!loggedInAsOrgCustomOrgUser.ManageUsers)
+                            {
+                                throw new BadRequestException("Your account is not permissioned to manage users.");
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1693]: please consider rewording, `"Your account does not have permission to manage users"`

File: src/Core/Utilities/CoreHelpers.cs [hunk 17]
```diff
@@ -719,6 +719,58 @@ public static X509Certificate2 GetIdentityServerCertificate(GlobalSettings globa
...
+                                if (org.ManagePolicies)
+                                {
+                                    claims.Add(new KeyValuePair<string, string>("managepolicies", org.Id.ToString()));
+                                }
+
+                                if (org.ManageUsers)
+                                {
+                                    claims.Add(new KeyValuePair<string, string>("manageusers", org.Id.ToString()));
+                                }
+
```
Comment (cscharf) [inline src/Core/Utilities/CoreHelpers.cs:771]: remove extra line

File: util/Migrator/DbScripts/2020-12-14_00_Permissions.sql [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+IF COL_LENGTH('[dbo].[OrganizationUser]', 'ManageUsers') IS NULL
+BEGIN
+    ALTER TABLE
+        [dbo].[OrganizationUser]
+    ADD
+        [ManageUsers] BIT NULL
+END
+GO
+
+UPDATE [dbo].[OrganizationUser]
```
Comment (cscharf) [inline util/Migrator/DbScripts/2020-12-14_00_Permissions.sql:82]: So, this is going to get expensive, however, you need to do the following:
1. Remove the `GO` statements between this `UPDATE` statement and the `ALTER COLUMN` statements
2. Wrap that whole block in a `TRANSACTION` so it "blocks" modification to prevent errors/issues with insertions happening before/after each of the statements

**Review by cscharf — COMMENTED** [review]
Hi @addisonbeck , great stuff. A few formatting nits, however one larger topic we should probably discuss around design approach in the DB tier.

File: bitwarden_license/src/Portal/Controllers/SsoController.cs (hunk 1 above)
Comment (addisonbeck) [reply bitwarden_license/src/Portal/Controllers/SsoController.cs]: My justification for this was tied to Policy people being able to enable/disable sso related polices, but there is room for distinction there. I added a `ManageSso`.

File: src/Core/Models/Table/OrganizationUser.cs (hunk 11 above)
Comment (addisonbeck) [reply src/Core/Models/Table/OrganizationUser.cs]: Implemented!

File: util/Migrator/DbScripts/2020-12-14_00_Permissions.sql (hunk 18 above)
Comment (addisonbeck) [reply util/Migrator/DbScripts/2020-12-14_00_Permissions.sql]: Another benefit of converting to a json blob: no updates to existing records need to be made anymore, we just do null checks and whatnot in the application logic. So, I guess this one is fixed.

File: src/Core/Utilities/CoreHelpers.cs [hunk 19]
```diff
@@ -778,5 +784,20 @@ public static X509Certificate2 GetIdentityServerCertificate(GlobalSettings globa
             }
             return claims;
         }
+
+        public static T LoadClassFromJsonData<T>(string jsonData) where T: new()
```
Comment (cscharf) [inline src/Core/Utilities/CoreHelpers.cs:788]: What whitespace after colon, `:` in generic type constraint

**Review by cscharf — COMMENTED** [review]
Please see comments and prior comments. Approach looks good.

File: src/Core/Services/Implementations/OrganizationService.cs (hunk 12 above)
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs:962]: If invite is null or there aren't any emails being invited we the invite method wouldn't be able to do any inviting, so I resolved this by just returning Not Found if either of those things are null.

File: src/Core/CurrentContext.cs (hunk 3 above)
Comment (addisonbeck) [reply src/Core/CurrentContext.cs]: Implemented with this. 

```
private Permissions SetOrganizationPermissionsFromClaims(string organizationId, Dictionary<string, IEnumerable<Claim>> claimsDict)
{
    bool hasClaim(string claimKey) 
    {
        return claimsDict.ContainsKey(claimKey) ? 
            claimsDict[claimKey].Any(x => x.Value == organizationId) : false;
    }

    return new Permissions
    {
        AccessBusinessPortal = hasClaim("accessbusinessportal"),
        AccessEventLogs = hasClaim("accesseventlogs"),
        AccessImportExport = hasClaim(""),
        AccessReports = hasClaim("accessreports"),
        ManageAllCollections = hasClaim(""),
        ManageAssignedCollections = hasClaim(""),
        ManageGroups = hasClaim(""),
        ManagePolicies = hasClaim(""),
        ManageSso = hasClaim("managesso"),
        ManageUsers = hasClaim("manageusers")
    };
}
```

File: src/Core/Models/Api/Request/Organizations/OrganizationUpdateRequestModel.cs (hunk 6 above)
Comment (addisonbeck) [reply src/Core/Models/Api/Request/Organizations/OrganizationUpdateRequestModel.cs]: Not applicable with IPermissions getting the boot

File: src/Core/Models/Business/OrganizationUserInvite.cs (hunk 7 above)
Comment (addisonbeck) [reply src/Core/Models/Business/OrganizationUserInvite.cs]: Not applicable with IPermissions getting the boot

File: src/Core/Models/Data/OrganizationUserOrganizationDetails.cs (hunk 9 above)
Comment (addisonbeck) [reply src/Core/Models/Data/OrganizationUserOrganizationDetails.cs]: Not applicable with IPermissions getting the boot

File: src/Core/Models/Interfaces/IPermissions.cs (hunk 10 above)
Comment (addisonbeck) [reply src/Core/Models/Interfaces/IPermissions.cs]: Not applicable with IPermissions getting the boot

File: src/Core/Models/Business/OrganizationUserInvite.cs (hunk 8 above)
Comment (cscharf) [reply src/Core/Models/Business/OrganizationUserInvite.cs:5]: Thanks! All sorted 😁 , however don't need the space or "groups", just overall sorting is fine.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 20]
```diff
@@ -1651,5 +1641,69 @@ private void ValidateOrganizationUpgradeParameters(Models.StaticStore.Plan plan,
...
+            };
+
+            if (checkedTypes.Contains(newType) || (oldType.HasValue && checkedTypes.Contains(oldType.Value)))
+            {
+                var loggedInUserOrgs = await _organizationUserRepository.GetManyByUserAsync(loggedInUserId);
+                var loggedInAsOrgOwner = loggedInUserOrgs
+                    .Any(u => u.OrganizationId == organizationId && u.Type == OrganizationUserType.Owner);
+                if (!loggedInAsOrgOwner)
+                {
+                    var isOwner = oldType.HasValue ?
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1661]: This can be simplified (and naturally handles `null` values):
```C#
var isOwner = oldType == OrganizationUserType.Owner;
```

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 21]
```diff
@@ -1651,5 +1641,69 @@ private void ValidateOrganizationUpgradeParameters(Models.StaticStore.Plan plan,
...
+                    var ownerUserConfigurationAttempt = (isOwner && nowOwner) || !(isOwner.Equals(nowOwner));
+                    if (ownerUserConfigurationAttempt)
+                    {
+                        throw new BadRequestException("Only an Owner can configure another Owner's account.");
+                    }
+
+                    var loggedInAsOrgAdmin = loggedInUserOrgs.Any(u => u.OrganizationId == organizationId && u.Type == OrganizationUserType.Admin);
+                    if (!loggedInAsOrgAdmin)
+                    {
+                        var isCustom = oldType.HasValue ?
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1674]: Can also be simplified, `var isCustom = oldType == OrganizationUserType.Custom`

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 22]
```diff
@@ -1651,5 +1641,69 @@ private void ValidateOrganizationUpgradeParameters(Models.StaticStore.Plan plan,
...
+                        var loggedInAsOrgCustomOrgUser = loggedInUserOrgs.FirstOrDefault(u => u.OrganizationId == organizationId && u.Type == OrganizationUserType.Custom);
+                        if (loggedInAsOrgCustomOrgUser != null)
+                        {
+                            var loggedInUserPermissions = CoreHelpers.LoadClassFromJsonData<Permissions>(loggedInAsOrgCustomOrgUser.Permissions);
+                            if (!loggedInUserPermissions.ManageUsers)
+                            {
+                                throw new BadRequestException("Your account does not have permission to manage users.");
+                            }
+
+                            var isAdmin = oldType.HasValue ?
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1693]: Simplify here as well

**Review by cscharf — COMMENTED** [review]
Final minor feedback.

File: src/Core/Models/Data/Permissions.cs [hunk 23]
```diff
@@ -0,0 +1,18 @@
+using System.Text.Json;
```
Comment (cscharf) [inline src/Core/Models/Data/Permissions.cs:1]: This `using` directive doesn't look necessary

File: src/Core/Services/IOrganizationService.cs [hunk 24]
```diff
@@ -8,6 +8,7 @@
 
 namespace Bit.Core.Services
 {
+    //hiiuguydjfkdfj
```
Comment (MGibson1) [inline src/Core/Services/IOrganizationService.cs:11]: Not sure if this comment gets its message across 😄

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 25]
```diff
@@ -1651,5 +1641,63 @@ private void ValidateOrganizationUpgradeParameters(Models.StaticStore.Plan plan,
...
+                OrganizationUserType.Admin,
+                OrganizationUserType.Custom, 
+            };
+
+            if (checkedTypes.Contains(newType) || (oldType.HasValue && checkedTypes.Contains(oldType.Value)))
+            {
+                var loggedInUserOrgs = await _organizationUserRepository.GetManyByUserAsync(loggedInUserId);
+                var loggedInAsOrgOwner = loggedInUserOrgs
+                    .Any(u => u.OrganizationId == organizationId && u.Type == OrganizationUserType.Owner);
+                if (!loggedInAsOrgOwner)
```
Comment (MGibson1) [inline src/Core/Services/Implementations/OrganizationService.cs:1659]: Personally, I'd invert the loggedInAs____ if statements to avoid this deep nesting

**Comment by addisonbeck** [comment]
@MGibson1 @cscharf tests are ready for review

File: test/Core.Test/Services/OrganizationServiceTests.cs [hunk 26]
```diff
@@ -138,5 +146,216 @@ public async Task OrgImportCreateNewUsersAndMarryExistingUser()
             await orgUserRepo.Received(1).UpsertAsync(Arg.Any<OrganizationUser>());
             await orgUserRepo.Received(2).CreateAsync(Arg.Any<OrganizationUser>());
         }
+
+        [Theory, CustomAutoDataAttribute(typeof(SutProviderCustomization))]
```
Comment (MGibson1) [inline test/Core.Test/Services/OrganizationServiceTests.cs:150]: not important, but you can remove the work `Attribute` from these references (keep Attribute in the class name)

File: test/Core.Test/Services/OrganizationServiceTests.cs [hunk 27]
```diff
@@ -138,5 +146,216 @@ public async Task OrgImportCreateNewUsersAndMarryExistingUser()
...
+            sutProvider.GetDependency<IOrganizationRepository>().GetByIdAsync(organizationId).Returns(Task.FromResult<Organization>(null));
+            var exception = await Assert.ThrowsAsync<NotFoundException>(
+                () => sutProvider.Sut.UpgradePlanAsync(organizationId, upgrade));
+        }
+
+        [Theory, CustomAutoDataAttribute(typeof(SutProviderCustomization))]
+        public async Task UpgradePlan_GatewayCustomIdIsNull_Throws(Organization organization, OrganizationUpgrade upgrade,
+                SutProvider<OrganizationService> sutProvider)
+        {
+            organization.GatewayCustomerId = string.Empty;
```
Comment (MGibson1) [inline test/Core.Test/Services/OrganizationServiceTests.cs:163]: Test name is that it's testing null, but this tests empty. Why not test both and send in the gatewayId as a paremeter?

Comment (MGibson1) [inline test/Core.Test/Services/OrganizationServiceTests.cs:163]: What if a GatewayCustomerId exists but is not valid for some reason (other than null or empty?)

File: test/Core.Test/AutoFixture/OrganizationFixtures.cs [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+            var validPlans = StaticStore.Plans.Where(p => !plansToIgnore.Contains(p.Type) && !p.Disabled).Select(p => p.Type).ToList();
+            fixture.Customize<OrganizationUpgrade>(composer => composer
+                .With(ou => ou.Plan, validPlans.Last()));
+            fixture.Customize<Organization>(composer => composer
+                .Without(o => o.GatewaySubscriptionId));
+        }
+    }
+    internal class OrganizationInvite : ICustomization
+    {
+        public OrganizationUserType InviteeUserType { get; set; }
```
Comment (MGibson1) [inline test/Core.Test/AutoFixture/OrganizationFixtures.cs:48]: Is the intended behavior to set these to null by default? If not, please set these properties as mandatory params in a constructor.

**Review by MGibson1 — COMMENTED** [review]
This looks good to me. I'm going to leave this review as just a comment to give @kspearrin and @cscharf chance for another go if they want it.

File: test/Core.Test/Services/OrganizationServiceTests.cs (hunk 27 above)
Comment (addisonbeck) [reply test/Core.Test/Services/OrganizationServiceTests.cs:163]: The check in the code is a NullOrWhitespace(). I know the tests should assume as little as possible about the code being tested, but does checking both in that context seem worth making a more complex test? In this case that would mean creating a whole new autodata attribute, wouldn't it?

Comment (addisonbeck) [reply test/Core.Test/Services/OrganizationServiceTests.cs:163]: There is no other validation checked on that field. An error throws if it doesn't exist, but as far as I can tell any other string is fair game here.

File: test/Core.Test/AutoFixture/OrganizationFixtures.cs (hunk 28 above)
Comment (addisonbeck) [reply test/Core.Test/AutoFixture/OrganizationFixtures.cs:48]: The intended behavior is to set these to `OranizationUserType.Owner` (via its int value `0`) by default. The attribute constructor specifies the default value for this in their init. The thought behind making these optional parameters with default values was to say "it doesn't matter what these values are" while also making sure AutoFixture isn't setting illogical or impossible combinations.

File: test/Core.Test/Services/OrganizationServiceTests.cs (hunk 26 above)
Comment (addisonbeck) [reply test/Core.Test/Services/OrganizationServiceTests.cs]: Done

File: test/Core.Test/Services/OrganizationServiceTests.cs (hunk 27 above)
Comment (MGibson1) [reply test/Core.Test/Services/OrganizationServiceTests.cs:162]: No, you could use InlineAutoData, unless that doesn't exist for organizations, but really I think this test is fine

Comment (MGibson1) [reply test/Core.Test/Services/OrganizationServiceTests.cs:162]: Is that the desired behavior? I don't know the code so maybe this isn't the part that should deal with it, but it seems like upgrade should be unsuccessful in some way if you can't actually charge for it

File: test/Core.Test/AutoFixture/OrganizationFixtures.cs (hunk 28 above)
Comment (MGibson1) [reply test/Core.Test/AutoFixture/OrganizationFixtures.cs:48]: I didn't realize the types were enums and thought they were classes. My mistake. I think the same question applies to blob, but that one makes more sense to default to null.

**Review by cscharf — APPROVED** [review]
Looks good @addisonbeck ! Thanks @MGibson1 for helping with the unit-tests, exciting stuff!
